Duration and context

andrewhare opened the pull request:
This PR does a few things:

* Changes the behavior of flows by allowing _sources_ to determine whether a duration is needed (that's why `--flow-duration` is now `--source-duration`).
* Get rid of `inputCtx` and `outputCtx` and use a single context for everything.
* Get rid of `--shutdown-duration` and have a single `--drain-duration` flag to control how log `vhs` continues to wait before shutting down the flow.

## New Behavior

Before this change, `vhs` would run for the allotted `--flow-duration` and then shut down the flow. This was ideal for capturing network traffic (since the different contexts would handle cancellation in a way that allowed data to continue to flow through until everything was saved) but it's not ideal for anything else.

The main problem is seen when you try to take network traffic in GCS (as an example, this would apply to any source) and run it back through `vhs`. What should the `--flow-duration` be set for on this run? The user has to estimate a length of time longer than the amount of time that it takes `vhs` to process the data, and even then users are only doing this to game the system since the `--flow-duration` doesn't have any meaning in this context so they are only supplying a value to keep `vhs` from shutting down in the middle of processing their data.

My solution is to change the semantics of how `vhs` thinks about sources and flows. I realized that the "duration" is really a property of a source, not the flow itself. So replace `--flow-duration` with `--source-duration`. This flag is now available for all sources that need it (e.g. `tcp.Source` needs it but `gcs.Source` does not).

So how does a flow decide when to shut down? When the input source closes its `Source.Streams()` channel. This close is propogated down through the input types until the `flow.Input` itself realizes that there are no more streams to process, and then it signals to the parent flow (via a new `done` channel) that there is no more data.

#### Implementation details about how source streams are tracked

There are two things that a `flow.Input` needs to keep track of:
1. Does `source.Streams()` have any more _streams_ of data to send me? An example of a source stream is a TCP connection, a GCS file, etc.
2. Has the last stream that was seen been read to `EOF`? This is accomplished by wrapping the reader in an `observe.ReadCloser` (new internal package and type) which checks for EOF and signals when `EOF` is reached. This allows the input to know that all input data has actually been read. Once the last stream is `EOF` we can safely say that all input has been processed.

@ztreinhart The `observe` package is a good place to start exploring the other types we talked about that allow us to know whats going on inside streams and channels.

## Contexts

This allowed me to remove 2 of the 3 contexts we've been using because now we only need one since we aren't using contexts to signal shutdown. We are using the natural close semantics of the channels to do this. I feel like this is a lot simpler and easier to work with.

This is definitely a draft and I'm hoping that we can use this as a jumping off point for further discussion.

ztreinhart reviewed (approved):
Nice work, I think this is a good improvement and it helps in handling a variety of inputs, as you noted in our meeting this morning. I wonder about going even further, though. What do you think about propagating a "finished" signal all the way through to the output sinks, so we could get rid of the drain duration, too? 

The drain duration seems like it will get trickier as we open this up and have a greater variety of modifiers, formats, sinks, etc. Setting an optimal drain will require some knowledge of the implementation of the components you're using, to account for internal buffers or other details that affect how long it takes for data to pass through the entire flow. Otherwise you just set what seems like a sufficiently long drain duration and hope for the best, which is basically what we've been doing with the flow duration...

Having some sort of finishing signal or end of data tag could also ensure deterministically that every bit of data from the source was actually processed, which could be valuable as well.

andrewhare commented:
I've addressed the test issues and made a few corrections. The most notable one is that we now need a drain duration on the input. It's not enough to close down the input on `EOF` like I hoped since slower systems will shut down before all data is passed from the input to the output. This is noticeable when running the tests in docker.

I think this is ready to be merged now with the latest fixes I added.

bradbeam commented on flow/output.go:
```diff
@@ -67,3 +77,29 @@ func (oo Outputs) Init(ctx session.Context) {
 }
+
+// Close closes all outputs.
+func (oo Outputs) Close(ctx session.Context) {
```
This name seems a bit misleading; maybe `Wait` would be more appropriate?

I was expecting this to do something like `close(o.Done())` or cancel the context but it does neither.

bradbeam replied:
Or is this to satisfy the `io.ReadCloser` interface?

andrewhare replied:
No, it doesn't satisfy an interface, its just a bad name for the function.

andrewhare replied:
@bradbeam: https://github.com/rename-this/vhs/pull/10/commits/c3107ff9ab331b82d7743a0d107744d153e574aa

bradbeam commented on internal/observe/read_closer.go:
```diff
@@ -0,0 +1,42 @@
+// ReadCloser is an observable io.ReadCloser.
+type ReadCloser interface {
+	io.ReadCloser
+	EOF() <-chan struct{}
```
Does this make sense to be a chan or should it just be a bool ( or maybe even reuse the io.EOF error type )?

andrewhare replied:
If we don't use a channel we'd have to poll the reader waiting for it to be at an `EOF` state. The channel allows us to respond to the `EOF` as an event.

bradbeam commented on tcp/stream_factory.go:
```diff
@@ -122,18 +126,29 @@ func (f *streamFactory) New(net, transport gopacket.Flow) tcpassembly.Stream {
 	}
 
-	f.out <- r
+	f.outMu.Lock()
+	if !f.closed {
+		f.out <- r
+	}
+	f.outMu.Unlock()
```
Should this be handled earlier in New?
```
func New() {
f.outMu.Lock()
defer f.outMu.Unlock()
if f.closed {
 return nil   ( or whatever makes sense here when the factory is closed )
}
...
f.out <- r
```
Stinks we'd hold the lock longer for each call, but it prevents us from doing a bunch of setup that is otherwise unnecessary.

andrewhare replied:
The `closed` state is only true when vhs is shutting down so I don't want to introduce more lock contention while vhs is actually running. This bool is in place to keep us from writing to the `out` channel once it has been closed so it makes the most sense to leave it as-is.
